### PR TITLE
Add various trait implementations for `Gc`

### PR DESCRIPTION
**Pull request by jacob-hughes**
This PR bundles a few semi-related commits together which would be a bit tedious to raise individually.

### CODE REVIEW

File: library/alloc/src/gc.rs [hunk 1]
```diff
@@ -295,6 +295,180 @@ impl<T> GcBox<MaybeUninit<T>> {
     }
 }
 
+impl<T: ?Sized + PartialEq> PartialEq for Gc<T> {
+    /// Equality for two `Gc`s.
+    ///
+    /// Two `Gc`s are equal if their inner values are equal, even if they are
+    /// stored in different allocation.
```
Comment (ltratt) [inline library/alloc/src/gc.rs:379]: s/allocation/allocations/

File: library/alloc/src/gc.rs [hunk 2]
```diff
@@ -257,6 +257,26 @@ impl<T> Gc<T> {
 }
 
 impl Gc<dyn Any> {
+    /// Attempt to downcast the `Gc<dyn Any>` to a concrete type.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #![feature(gc)]
```
Comment (ltratt) [inline library/alloc/src/gc.rs:266]: Is the `# #!` bit deliberate (i.e. two `#`s)?

Comment (jacob-hughes) [reply library/alloc/src/gc.rs:266]: Yes, the first `#` means that the line is not rendered in the docs example but it's needed for the example to compile. It's a way to elide setup code that make the examples harder to read.

Comment (ltratt) [reply library/alloc/src/gc.rs:266]: Aha!

File: library/alloc/src/gc.rs (hunk 1 above)
Comment (jacob-hughes) [reply library/alloc/src/gc.rs]: Fixed: ae978a21b8dd3f3fd771f5270af975ca132a121c

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**